clicore/logger.go

```diff
@@ -0,0 +1,165 @@
+// - A human-friendly console logger that writes info and warning messages to stderr
+func BuildCliLogger(pathElements []string, outputModeEnvVar string) *zap.SugaredLogger {
+	return buildCliLoggerOptions(pathElements, outputModeEnvVar, nil)
+}
+
+//BuildMockedCliLogger is the test-environment counterpart of BuildCliLogger
+// It stores log output in buffers that can be inspected by tests.
+func BuildMockedCliLogger(pathElements []string, outputModeEnvVar string, mockTargets *MockTargets) *zap.SugaredLogger {
+	return buildCliLoggerOptions(pathElements, outputModeEnvVar, mockTargets)
+}
+
+func buildCliLoggerOptions(pathElements []string, outputModeEnvVar string, mockTargets *MockTargets) *zap.SugaredLogger {
+	verboseMode := os.Getenv(outputModeEnvVar) == "1"
+	fileCore := buildCliZapCoreFile(pathElements, verboseMode, mockTargets)
+	consoleCores := buildCliZapCoreConsoles(verboseMode, mockTargets)
+	allCores := consoleCores
+	if fileCore != nil {
+		allCores = append(allCores, fileCore)
+	}
+	core := zapcore.NewTee(allCores...)
+	logger := zap.New(core).Sugar()
+	return logger
+}
+
+//FilePathFromHomeDir is a utility that makes it easier to find the absolute path to a file, given its file path
+// elements relative to its home directory.
+// pathElementsRelativeToHome is passed as an array to avoid os-specific directory delimiter complications
+// example: []string{".config","default.yaml"}
+func FilePathFromHomeDir(pathElementsRelativeToHome []string) (string, error) {
+	// Find home directory.
+	home, err := homedir.Dir()
+	if err != nil {
+		return "", err
+	}
+	nPathElements := len(pathElementsRelativeToHome)
+	nDirElements := nPathElements - 1
+	if nDirElements > 0 {
+		dirsToMake := append([]string{home}, pathElementsRelativeToHome[:nDirElements]...)
+		// if the directory already exists, MkdirAll returns nil
+		if err := os.MkdirAll(filepath.Join(dirsToMake...), 0755); err != nil {
+			return "", err
+		}
+	}
+	pathElements := append([]string{home}, pathElementsRelativeToHome...)
+	return filepath.Join(pathElements...), nil
+}
+
+func buildCliZapCoreFile(pathElements []string, verboseMode bool, mockTargets *MockTargets) zapcore.Core {
+	// force to verbose mode while running in a test environment
+	if mockTargets != nil {
+		verboseMode = true
+	}
+	path, err := FilePathFromHomeDir(pathElements)
+	if err != nil {
+		if verboseMode {
+			// we don't want to return errors just because we cannot write logs to a file
+			// users can use the verbose flag to get full output to the console
+			fmt.Printf("Could not open log file %s for writing: %v\n", filepath.Join(pathElements...), err)
+		}
+		return nil
+	}
+
+	// if we decide we want to append logs, we can do it this way:
+	// (we would be required to first create the file and "rotate" it as it grows)
+	//file, err := os.OpenFile(path, os.O_WRONLY|os.O_APPEND, os.ModeAppend)
+
+	// for now, let's just create/overwrite the file each time
+	file, err := os.Create(path)
```

Comment:
fwiw this line will open a file in append mode if it exists, or create it if if doesn't

`file, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)`

Reply:
Interesting, I wonder if `O_APPEND` is a superset of `O_CREATE` because the line I referenced worked too. Either way, I think for now we don't want to risk growing massive log files so storing only the output from the last command seems right to me.